src/data_types/tensor.hpp

```diff
@@ -0,0 +1,432 @@
+    }
+
+private:
+    static_assert(rank() > 0);
+
+public:
+    /**
+     * @brief Construct an uninitialised tensor object.
+     */
+    explicit KOKKOS_DEFAULTED_FUNCTION Tensor() = default;
+
+    /**
+     * @brief Construct a tensor object initialised with a value.
+     * @param fill_value The value with which the tensor should be filled.
+     */
+    explicit KOKKOS_FUNCTION Tensor(ElementType fill_value)
+    {
+        m_data.fill(fill_value);
+    }
+
+    /**
+     * @brief Construct a 1D tensor object by providing the elements that should
+     * be saved in the vector.
+     *
+     * @param elements The elements of the tensor.
+     */
+    template <
+            class... Params,
+            class = std::enable_if_t<(std::is_convertible_v<Params, ElementType> && ...)>,
+            class = std::enable_if_t<sizeof...(Params) == size() && sizeof...(Params) != 1>>
+    explicit KOKKOS_FUNCTION Tensor(Params... elements) : m_data({elements...})
+    {
+        static_assert(
+                rank() == 1,
+                "Filling the tensor on initialisation is only permitted for 1D vector objects");
+    }
+
+    /**
+     * @brief Construct a tensor object by copying an existing tensor.
+     *
+     * @param o_tensor The tensor to be copied.
+     */
+    template <class OElementType>
+    explicit KOKKOS_FUNCTION Tensor(Tensor<OElementType, ValidIndexSet...> const& o_tensor)
+        : m_data(o_tensor.m_data)
+    {
+    }
+
+    /**
+     * @brief Get a modifiable reference to an element of the tensor.
+     * @tparam QueryIndexTag A type describing the relevant index.
+     * @return The relevant element of the tensor.
+     */
+    template <class QueryTensorIndexElement>
+    KOKKOS_FUNCTION ElementType& get()
+    {
+        static_assert(tensor_tools::is_tensor_index_element_v<QueryTensorIndexElement>);
+        return m_data[QueryTensorIndexElement::index()];
+    }
+
+    /**
+     * @brief Get an element of the tensor.
+     * @tparam QueryIndexTag A type describing the relevant index.
+     * @return The relevant element of the tensor.
+     */
+    template <class QueryTensorIndexElement>
+    KOKKOS_FUNCTION ElementType const& get() const
+    {
+        static_assert(tensor_tools::is_tensor_index_element_v<QueryTensorIndexElement>);
+        return m_data[QueryTensorIndexElement::index()];
+    }
+
+    /**
+     * @brief A copy operator.
+     * @param other The tensor to be copied.
+     * @return A reference to the current tensor.
+     */
+    KOKKOS_DEFAULTED_FUNCTION Tensor& operator=(Tensor const& other) = default;
+
+    /**
+     * @brief An operator to multiply all the element of the current tensor by
+     * a value.
+     * @param val The value by which the elements should be multiplied.
+     * @return A reference to the current modified tensor.
+     */
+    template <class OElementType>
+    KOKKOS_FUNCTION Tensor& operator*=(OElementType val)
+    {
+        for (int i(0); i < s_n_elements; ++i) {
```

Comment:
```suggestion
        for (std::size_t i(0); i < s_n_elements; ++i) {
```

`s_n_elements` is of type `std::size_t`.

Same for other operators.

Reply:
Fixed [9862af6](https://github.com/gyselax/gyselalibxx/pull/81/commits/9862af662259f1c8fa4a403326e9a725ef213b03)